ledger/eval.go

```diff
@@ -682,20 +683,24 @@ func (eval *BlockEvaluator) prepareEvalParams(txgroup []transactions.SignedTxnWi
 		// Initialize side effects and group without ApplyData lazily
 		if groupNoAD == nil {
 			groupNoAD = make([]transactions.SignedTxn, len(txgroup))
-			for j := range txgroup {
+			for j, tx := range txgroup {
 				groupNoAD[j] = txgroup[j].SignedTxn
+				if tx.SignedTxn.Txn.Type == protocol.ApplicationCallTx && eval.proto.EnableAppFeePooling {
+					pooledApplicationBudget += eval.proto.MaxAppProgramCost
+				}
 			}
 			pastSideEffects = logic.MakePastSideEffects(len(txgroup))
 			minTealVersion = logic.ComputeMinTealVersion(groupNoAD)
 		}
 
 		res[i] = &logic.EvalParams{
-			Txn:             &groupNoAD[i],
-			Proto:           &eval.proto,
-			TxnGroup:        groupNoAD,
-			GroupIndex:      i,
-			PastSideEffects: pastSideEffects,
-			MinTealVersion:  &minTealVersion,
+			Txn:                     &groupNoAD[i],
+			Proto:                   &eval.proto,
+			TxnGroup:                groupNoAD,
+			GroupIndex:              i,
+			PastSideEffects:         pastSideEffects,
+			MinTealVersion:          &minTealVersion,
+			PooledApplicationBudget: pooledApplicationBudget,
```

Comment:
I think your life gets way, way simpler if you just make this a pointer, so you shared &pooledApplicationBudget in all eps.  Then each one decrements as it goes.  If you ever hit zero, you're done then.

Then (at least by line count) your patch gets much much smaller too, because all of the interfaces stay the same.

Reply:
I think I like that suggestion -- I'll take a look at your other comments later, but I'll work on this first.